fix #8871 runnableExamples now preserves source code comments, litterals, and all formatting; other bug fix

timotheecour opened the pull request:
* fix #8871
* EDIT: fix #9345
* fix a regression introduced in https://github.com/nim-lang/Nim/pull/14441 that would cause skipping runnableExamples in templates in some cases; and made the logic cleaner (and easier to change if needed) by using a simple state machine instead of a recursive call in `getAllRunnableExamples`
* fix https://github.com/nim-lang/Nim/issues/14473 (similar/same as above point)
* found a corner from 1st version of this PR and fixed it, see `testNimDocTrailingExample`, because `numLines` is inclusive

most imporantly:
* add a lot of regression tests to prevent future regressions


so now both #8871 and #8871 as can be seen in by comparing `nim doc --outdir:htmldocs --project nimdoc/testproject/testproject.nim`:


## for #8871
right before https://github.com/nim-lang/Nim/pull/14441:
![image](https://user-images.githubusercontent.com/2194784/82994007-5492ea00-9fb6-11ea-9416-3392ccf33b59.png)

after this PR:
![image](https://user-images.githubusercontent.com/2194784/82993941-3e852980-9fb6-11ea-974c-6dd2847d7153.png)

=> we're now preserving formatting exactly (in particular we don't skip comments, change litterals and change indentation etc)


## and for #8871
right before https://github.com/nim-lang/Nim/pull/14441:
![image](https://user-images.githubusercontent.com/2194784/82994195-9cb20c80-9fb6-11ea-92ec-a397b7511c91.png)


after this PR:
![image](https://user-images.githubusercontent.com/2194784/82994090-755b3f80-9fb6-11ea-92f5-45fb04008f57.png)

=> we're now not skipping doc comments (C) after the 1st runnableExamples (R) so if you have CRCRC you end up with CRCRC instead of CRR; also we're now preserving ordering

timotheecour committed:
fix #8871 runnableExamples now preserves source code comments, litterals, and all formatting

kaushalmodi commented:
This is awesome! Just yesterday, I noticed that runnableExamples was reformatting my manually set indentation; e.g. [this][1] rendered to https://kaushalmodi.github.io/version/#getVersion%2CopenArray%5Bstring%5D%2Cstring

[1]: https://github.com/kaushalmodi/version/blob/09100e7d619e8b681bb794d7b38d2b66e23a8af9/src/version.nim#L139-L147

Araq commented:
Did you consider to re-use nimpretty's approach? Also, is it time to make ``nimdoc`` a tool of its own?

timotheecour committed:
remove orig deadcode from getAllRunnableExamplesImpl

timotheecour committed:
fix expected examples

timotheecour commented:
@araq PTAL (you can look at generated html for `nim doc --project -r nimdoc/testproject/testproject.nim`)

> Did you consider to re-use nimpretty's approach?

nimpretty doesn't do syntax highlighting so the code I wrote is still needed even with reusing/calling nimpretty.
Future work could nimpretty-fy the source code string (it may or may not be desirable though eg wrt preserving user formatted examples). 

This PR got a bit larger than initially because I fixed other issues (see updated note in top post) and added a lot of tests, but the part that's responsible for extracting and rendering runnableExampes (ie fixing #8871) is actually quite concise.

> Also, is it time to make nimdoc a tool of its own?

IMO we should completely separate the reflection part (getting postprocessed (after semcheck) AST with module infos, symbols, doc comments properly attached, runnableExamples-as-string, etc) from the rendering part (including running runnableExamples which calls nim-as-binary and doesn't need to depend on compiler sources). All rendering code should be separated from the compiler and would just consume what the 1st step produces (IR, json, PNode, etc see below though).

and here's how it should work IMO:

## macros should be powerful enough to enable full reflection and 100% replace docgen

We can have really clean concise `nim-doc-as-library` by adding just a few magics that would enable full reflection (representing AST after semcheck for each module as 1 NimNode (typed AST)), **without even having to `import compiler`**.

This is one of the uses cases I had in mind with https://github.com/nim-lang/Nim/pull/9560 and https://github.com/nim-lang/Nim/pull/8903 and I actually had a working prototype of `nim-doc` in pure user code using those 2 PR's, that did not need `import compiler`. It didn't handle many important things like RST parsing, running runnableExamples etc from doc comments, but would be doable especially if we separate rst-logic from compiler and usable in a separate module (stdlib or fusion, probably stdlib since compiler still needs it for now )

Note: this approach invovles 0 serialization (no IR or json needed) since the macros get all they need via a NimNode for each module representing all symbols. It's then possible for user code to then convert these to whatever format needed (json, etc) or to re-do what nim doc does.

## other use cases for reflection beyond nim doc
these could be implemented in user code via macros by exposing AST after semantic phase:
* search tool (eg: list all routines that take an `openArray[string]` as 1st argument
* linter (eg: report when `result` is un-assigned)
* nimfix (eg: replace noop code like `myseq = @[]` initializations)
etc

Clyybber commented:
@timotheecour Love the idea of docgen as a macro.
Maybe it can be done without additional magics via this approach:
```nim
import macros

macro genDocs(e: typed) =
  echo treeRepr e

genDocs:
  include moduleWeWantToGenerateDocsFor
```

Araq commented:
> macros should be powerful enough to enable full reflection and 100% replace docgen

I completely disagree. This idea that we can never ever *reduce* an AST because it might be rendered back to ascii somewhere did cost us years of implementation effort, keeps costing us effort and makes writing macros more complex. What's the benefit? A good docgen is still as hard to implement as it was before, if not harder and you'll put pressure on the Nim VM.

kaushalmodi commented:
@timotheecour Referring to your comment in https://github.com/nim-lang/Nim/issues/14473#issuecomment-634927880,

> this should be fixed in #14439 (I just tried), can you confirm on your end and also for your project that #14439 works?

Yes, I just confirmed that this PR fixes the minimal example in https://github.com/nim-lang/Nim/issues/14473, as well as in my real project.

Can you update a test case in this PR to also `import std/[sequtils]` and use one of the exported symbols from that in the test case to match the test case in https://github.com/nim-lang/Nim/issues/14473 ?

Thanks!

timotheecour committed:
add test to close https://github.com/nim-lang/Nim/issues/14473

timotheecour commented:
> Can you update a test case in this PR to also import std/[sequtils] and use one of the exported symbols from that in the test case to match the test case in #14473 ?

done. (see also: https://github.com/nim-lang/RFCs/issues/231)

timotheecour commented:
friendly ping @araq because this fixes a very recent regression that I've introduced in  #14441 that prevents some runnableExamples from running; let me know if you instead want me
* A to split it in 2 PR's but, since this PR adds a number of tests (which I should've done in https://github.com/nim-lang/Nim/pull/14441 ...) it should give more confidence
* B revert https://github.com/nim-lang/Nim/pull/14441 in the meantime

Araq commented on compiler/docgen.nim:
```diff
@@ -529,67 +534,104 @@ proc prepareExample(d: PDoc; n: PNode): string =
+
+proc getRoutineBody(n: PNode): PNode =
+  ## works around a transf bug D20200526T163511 where `result` messes AST
+  # TODO: try w templates and macro, see if they're affected
```
What's wrong  with ``bodyPos``?

timotheecour replied:
will reply below so this doesn't get burried

Araq commented on compiler/renderverbatim.nim:
```diff
@@ -0,0 +1,65 @@
+  for j in 0..<indent:
+    if src.len <= j: return true
+    if src[j] == ' ': continue
+    return false
```
Most convoluted control flow. ``if src[j] != ' ': return false``.

timotheecour replied:
done

Araq commented on compiler/renderverbatim.nim:
```diff
@@ -0,0 +1,65 @@
+  ## we'd need to check performance impact to enable it for nimdoc.
+  let first = n.lastSon.info
+  let last = n.lastNodeRec.info
+  var ret = ""
```
Use ``result``

timotheecour replied:
done

Araq commented on compiler/renderverbatim.nim:
```diff
@@ -0,0 +1,65 @@
+
+proc extractRunnableExamplesSource*(conf: ConfigRef; n: PNode): string =
+  ## TLineInfo.offsetA,offsetB would be cleaner but it's only enabled for nimpretty,
+  ## we'd need to check performance impact to enable it for nimdoc.
```
Not a fan, but ok.

timotheecour replied:
it'd avoid some hacks I'm doing, but the hacks aren't too bad and very localized

Araq commented on compiler/renderverbatim.nim:
```diff
@@ -0,0 +1,65 @@
+  ## TLineInfo.offsetA,offsetB would be cleaner but it's only enabled for nimpretty,
+  ## we'd need to check performance impact to enable it for nimdoc.
+  let first = n.lastSon.info
+  let last = n.lastNodeRec.info
```
But what if the "last" node is the result of a template expansion and refers to a different file entirely?

timotheecour replied:
that can't happen AFAIK because template isn't semchecked at this point; see the `fromUtilsGen` test I added and look at the resulting html

timotheecour committed:
fix comments

timotheecour committed:
fix test

timotheecour committed:
improve comment

timotheecour committed:
correctly handle regular comments before 1st token inside runnableExamples

timotheecour committed:
add test to answer https://github.com/nim-lang/Nim/pull/14439#discussion_r431829199

timotheecour committed:
update tests

timotheecour commented:
@araq PTAL; all points addressed and found I more corner case that i fixed (regular comments before the 1st node inside a runnableExamples are now correctly handled, ie appear as part of the doc comment)


in reply to https://github.com/nim-lang/Nim/pull/14439#discussion_r431822781

> What's wrong with bodyPos?

I added some explanation inside PR code; whether it's a bug or not I'm not sure, but it seems inconsistent that you have:
```nim
  proc someType*(): int =
    ## foo
    result = 3
=> this is weird
  result =
    ## foo
    3;

  proc someType*(): int =
    ## foo
    3
=> this is fine
  ## foo
  result = 3;
```

IMO the sane behavior is for transf to transform the 1st case into the following instead:
```nim
## foo
result = 3;
```


This has very strong echoes of https://github.com/nim-lang/Nim/issues/14420#issuecomment-633711557

EDIT: simplified further in https://github.com/nim-lang/Nim/pull/14701
